csi/src/block_vol.rs

```diff
@@ -49,21 +53,41 @@ pub async fn publish_block_volume(
             error
         )
     })? {
-        let devt = unsafe { libc::makedev(259, 254) };
-
-        let cstr_dst = std::ffi::CString::new(target_path.as_str()).unwrap();
-        let res =
-            unsafe { libc::mknod(cstr_dst.as_ptr(), libc::S_IFBLK, devt) };
-
-        if res != 0 {
-            return Err(failure!(
-                Code::Internal,
-                "Failed to publish volume {}: mknod at {} failed",
-                volume_id,
-                target_path
-            ));
+        // Idempotency, if we have done this already just return success.
+        match findmnt_device(target_path) {
```

Comment:
I think that we should only perform this check if target_path actually exists - and then only if it is a device special and not a regular file. Also, we should create the file only if it did not exist previously - rather than implicitly truncating an existing file. If the file exists already and has size 0 we can just bind mount it. If it exists and has non-zero size we should probably abort as truncating such a file means potential data loss.